src/lib/ast.rs

```diff
@@ -0,0 +1,623 @@
+
+use std::borrow::Cow;
+use std::collections::HashMap;
+use std::convert::{TryFrom, TryInto};
+use std::fmt;
+use std::fs::File;
+use std::io;
+use std::io::{Read, Write};
+use std::ops::{Index, IndexMut};
+use std::path::Path;
+
+use dot;
+use lrlex::{build_lex, Lexeme};
+use lrpar::parser;
+use lrtable::{Grammar, Minimiser, TIdx, yacc_to_statetable};
+
+#[derive(Debug)]
+/// Errors raised when parsing a source file.
+pub enum ParseError {
+    /// Io error returned from standard library routine.
+    Io(io::Error),
+    /// File not found.
+    FileNotFound,
+    /// File name had no extension (used to determine which lexer/parser to use).
+    NoFileExtension,
+    /// Lexer could not be built by `lrlex`.
+    BrokenLexer,
+    /// Parser could not be built by `lrpar`.
+    BrokenParser,
+    /// File contained lexical error and could not be lexed.
+    LexicalError,
+    /// File contained syntax error and could not be parsed.
+    SyntaxError,
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq, PartialOrd)]
+/// A node identifier for nodes within a particular `Arena`.
+pub struct NodeId {
+    index: usize,
+}
+
+/// A type to represent edges between nodes within an `Arena`.
+type EdgeId = (usize, usize);
+
+impl fmt::Display for NodeId {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{:}", self.index)
+    }
+}
+
+/// An arena in which to place `NodeID`-indexed AST nodes.
+pub struct Arena<T, U> {
+    nodes: Vec<Node<T, U>>,
+}
+
+impl<T, U> Index<NodeId> for Arena<T, U> {
+    type Output = Node<T, U>;
+
+    fn index(&self, node: NodeId) -> &Node<T, U> {
+        &self.nodes[node.index]
+    }
+}
+
+impl<T, U> IndexMut<NodeId> for Arena<T, U> {
+    fn index_mut(&mut self, node: NodeId) -> &mut Node<T, U> {
+        &mut self.nodes[node.index]
+    }
+}
+
+impl<T, U> Arena<T, U> {
+    /// Create an empty `Arena`.
+    pub fn new() -> Arena<T, U> {
+        Arena { nodes: Vec::new() }
+    }
+
+    /// Create a new node from its data.
+    pub fn new_node(&mut self, data: T, ty: U, indent: u16) -> NodeId {
+        let next_index = self.nodes.len();
+        let node = Node::new(data, ty, indent);
+        self.nodes.push(node);
+        NodeId { index: next_index }
+    }
+
+    /// Return `true` if the arena is empty, `false` otherwise.
+    pub fn is_empty(&self) -> bool {
+        self.nodes.is_empty()
+    }
+
+    /// Make one node the child of another.
+    /// Panic if either `NodeId` is not found or this arena is empty.
+    pub fn make_child_of(&mut self, child: NodeId, parent: NodeId) {
+        if child.index >= self.nodes.len() {
+            panic!("Node ID {} not found in arena.", child);
+        }
+        if parent.index >= self.nodes.len() {
+            panic!("Node ID {} not found in arena.", parent);
+        }
+        if self.nodes.is_empty() {
+            panic!("Arena is empty.")
+        }
+        self[child].parent = Some(parent);
+        match self[parent].first_child {
```

Comment:
Here we are making sure that the parent node lists the new node as one of its children. Either as the first and last child (if the parent node does not have any children yet) or as the last child (if it does already have one or more children).